File: apps/executable-spec-web/src/platform/state_machine/transitions.ts
export type UnifiedExecutionState =
  | "INIT"
  | "READY"
  | "RUNNING"
  | "BLOCKED"
  | "PASSED"
  | "FAILED"
  | "SIGNING"
  | "ARCHIVED";

export type StateScope = "NODE" | "CONTAINER";

export const STATE_SCOPE_NODE: StateScope = "NODE";
export const STATE_SCOPE_CONTAINER: StateScope = "CONTAINER";

export const ERROR_UNKNOWN_SCOPE = "SM_STATE_UNKNOWN_SCOPE";
export const ERROR_UNKNOWN_STATUS = "SM_STATE_UNKNOWN_STATUS";
export const ERROR_ILLEGAL_TRANSITION = "SM_STATE_ILLEGAL_TRANSITION";

export type StateTransitionErrorCode =
  | typeof ERROR_UNKNOWN_SCOPE
  | typeof ERROR_UNKNOWN_STATUS
  | typeof ERROR_ILLEGAL_TRANSITION;

export interface StateTransitionErrorPayload {
  code: StateTransitionErrorCode;
  message: string;
  scope?: string;
  current?: string;
  target?: string;
  allowed?: string[];
}

export class StateTransitionError extends Error {
  readonly code: StateTransitionErrorCode;
  readonly scope?: string;
  readonly current?: string;
  readonly target?: string;
  readonly allowed: string[];

  constructor(payload: StateTransitionErrorPayload) {
    super(payload.message);
    this.name = "StateTransitionError";
    this.code = payload.code;
    this.scope = payload.scope;
    this.current = payload.current;
    this.target = payload.target;
    this.allowed = [...(payload.allowed ?? [])];
  }

  asPayload(): StateTransitionErrorPayload {
    return {
      code: this.code,
      message: this.message,
      scope: this.scope,
      current: this.current,
      target: this.target,
      allowed: [...this.allowed],
    };
  }
}

const UNIFIED_STATES: readonly UnifiedExecutionState[] = [
  "INIT",
  "READY",
  "RUNNING",
  "BLOCKED",
  "PASSED",
  "FAILED",
  "SIGNING",
  "ARCHIVED",
] as const;

const LEGACY_STATE_ALIASES: Readonly<Record<string, UnifiedExecutionState>> = {
  INIT: "INIT",
  READY: "READY",
  RUNNING: "RUNNING",
  BLOCKED: "BLOCKED",
  PASSED: "PASSED",
  FAILED: "FAILED",
  SIGNING: "SIGNING",
  ARCHIVED: "ARCHIVED",
  DRAFT: "INIT",
  PENDING: "INIT",
  UNLOCKED: "READY",
  IN_PROGRESS: "RUNNING",
  COMPUTED: "RUNNING",
  GATED: "RUNNING",
  GATE_PASS: "SIGNING",
  GATE_FAIL: "FAILED",
  PASS: "PASSED",
  FINAL_PASS: "PASSED",
  QUALIFIED: "PASSED",
  VALIDATED: "PASSED",
  VERIFIED: "PASSED",
  COMPLETED: "PASSED",
  SUCCESS: "PASSED",
  OVERRIDDEN: "PASSED",
  FAIL: "FAILED",
  FINAL_FAIL: "FAILED",
  REJECTED: "FAILED",
  CRITICAL: "FAILED",
  ERROR: "FAILED",
  LOCKED: "BLOCKED",
};

export const PAGE_TEXT_STATE_MAP_ZH_CN: Readonly<Record<string, UnifiedExecutionState>> = {
  草稿: "INIT",
  就绪: "READY",
  可执行: "READY",
  执行中: "RUNNING",
  进行中: "RUNNING",
  阻塞: "BLOCKED",
  受阻: "BLOCKED",
  已阻断: "BLOCKED",
  通过: "PASSED",
  已完成: "PASSED",
  已验证: "PASSED",
  合格: "PASSED",
  不通过: "FAILED",
  未通过: "FAILED",
  已驳回: "FAILED",
  签名中: "SIGNING",
  已归档: "ARCHIVED",
};

const NODE_TRANSITIONS: Readonly<Record<UnifiedExecutionState, readonly UnifiedExecutionState[]>> = {
  INIT: ["READY", "RUNNING", "BLOCKED"],
  READY: ["RUNNING", "BLOCKED", "ARCHIVED"],
  RUNNING: ["PASSED", "FAILED", "BLOCKED", "SIGNING"],
  BLOCKED: ["READY", "RUNNING", "FAILED"],
  PASSED: ["SIGNING", "ARCHIVED"],
  FAILED: ["READY", "RUNNING", "SIGNING", "ARCHIVED"],
  SIGNING: ["PASSED", "FAILED", "ARCHIVED"],
  ARCHIVED: [],
};

const CONTAINER_TRANSITIONS: Readonly<Record<UnifiedExecutionState, readonly UnifiedExecutionState[]>> = {
  INIT: ["READY", "RUNNING", "BLOCKED"],
  READY: ["RUNNING", "BLOCKED", "ARCHIVED"],
  RUNNING: ["READY", "BLOCKED", "PASSED", "FAILED", "SIGNING"],
  BLOCKED: ["READY", "RUNNING", "FAILED"],
  PASSED: ["SIGNING", "ARCHIVED"],
  FAILED: ["READY", "RUNNING", "SIGNING", "ARCHIVED"],
  SIGNING: ["PASSED", "FAILED", "ARCHIVED"],
  ARCHIVED: [],
};

const TRANSITIONS_BY_SCOPE: Readonly<Record<StateScope, Readonly<Record<UnifiedExecutionState, readonly UnifiedExecutionState[]>>>> = {
  NODE: NODE_TRANSITIONS,
  CONTAINER: CONTAINER_TRANSITIONS,
};

function normalizeToken(value: unknown): string {
  return String(value ?? "").trim().toUpperCase();
}

function normalizeScope(scope: string): StateScope {
  const token = normalizeToken(scope);
  if (token === STATE_SCOPE_NODE) {
    return STATE_SCOPE_NODE;
  }
  if (token === STATE_SCOPE_CONTAINER) {
    return STATE_SCOPE_CONTAINER;
  }
  throw new StateTransitionError({
    code: ERROR_UNKNOWN_SCOPE,
    message: `Unknown state machine scope: ${scope}`,
    scope,
  });
}

export function normalizeState(value: unknown, scope?: string): UnifiedExecutionState {
  if (typeof scope !== "undefined") {
    normalizeScope(scope);
  }
  const token = normalizeToken(value);
  const mapped = LEGACY_STATE_ALIASES[token];
  if (mapped) {
    return mapped;
  }
  throw new StateTransitionError({
    code: ERROR_UNKNOWN_STATUS,
    message: `Unknown status value: ${String(value)}`,
    scope,
    current: String(value ?? ""),
  });
}

export function normalizePageStatusText(value: unknown): UnifiedExecutionState {
  const text = String(value ?? "").trim();
  if (!text) {
    throw new StateTransitionError({
      code: ERROR_UNKNOWN_STATUS,
      message: "Page status text is empty.",
      current: "",
    });
  }
  const mapped = PAGE_TEXT_STATE_MAP_ZH_CN[text];
  if (mapped) {
    return mapped;
  }
  return normalizeState(text);
}

export function allowedTargets(scope: string, currentState: unknown): UnifiedExecutionState[] {
  const normalizedScope = normalizeScope(scope);
  const normalizedCurrent = normalizeState(currentState, normalizedScope);
  return [...TRANSITIONS_BY_SCOPE[normalizedScope][normalizedCurrent]];
}

export function canTransition(scope: string, currentState: unknown, targetState: unknown): boolean {
  const normalizedScope = normalizeScope(scope);
  const normalizedCurrent = normalizeState(currentState, normalizedScope);
  const normalizedTarget = normalizeState(targetState, normalizedScope);
  if (normalizedCurrent === normalizedTarget) {
    return true;
  }
  return TRANSITIONS_BY_SCOPE[normalizedScope][normalizedCurrent].includes(normalizedTarget);
}

export function assertTransition(scope: string, currentState: unknown, targetState: unknown): UnifiedExecutionState {
  const normalizedScope = normalizeScope(scope);
  const normalizedCurrent = normalizeState(currentState, normalizedScope);
  const normalizedTarget = normalizeState(targetState, normalizedScope);
  if (normalizedCurrent === normalizedTarget) {
    return normalizedTarget;
  }
  const allowed = TRANSITIONS_BY_SCOPE[normalizedScope][normalizedCurrent];
  if (!allowed.includes(normalizedTarget)) {
    throw new StateTransitionError({
      code: ERROR_ILLEGAL_TRANSITION,
      message: `Illegal transition: ${normalizedScope} ${normalizedCurrent} -> ${normalizedTarget}`,
      scope: normalizedScope,
      current: normalizedCurrent,
      target: normalizedTarget,
      allowed: [...allowed],
    });
  }
  return normalizedTarget;
}

export function transition(scope: string, currentState: unknown, targetState: unknown): UnifiedExecutionState {
  return assertTransition(scope, currentState, targetState);
}

export function toSchedulerTaskStatus(state: UnifiedExecutionState): "ready" | "blocked" | "running" | "pass" | "failed" {
  if (state === "RUNNING" || state === "SIGNING") {
    return "running";
  }
  if (state === "BLOCKED") {
    return "blocked";
  }
  if (state === "PASSED" || state === "ARCHIVED") {
    return "pass";
  }
  if (state === "FAILED") {
    return "failed";
  }
  return "ready";
}

export function listUnifiedStates(): UnifiedExecutionState[] {
  return [...UNIFIED_STATES];
}

File: backend/app/state_machine/transitions.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Final, Literal

UnifiedState = Literal[
    "INIT",
    "READY",
    "RUNNING",
    "BLOCKED",
    "PASSED",
    "FAILED",
    "SIGNING",
    "ARCHIVED",
]
StateScope = Literal["NODE", "CONTAINER"]

STATE_INIT: Final[UnifiedState] = "INIT"
STATE_READY: Final[UnifiedState] = "READY"
STATE_RUNNING: Final[UnifiedState] = "RUNNING"
STATE_BLOCKED: Final[UnifiedState] = "BLOCKED"
STATE_PASSED: Final[UnifiedState] = "PASSED"
STATE_FAILED: Final[UnifiedState] = "FAILED"
STATE_SIGNING: Final[UnifiedState] = "SIGNING"
STATE_ARCHIVED: Final[UnifiedState] = "ARCHIVED"

UNIFIED_STATES: Final[tuple[UnifiedState, ...]] = (
    STATE_INIT,
    STATE_READY,
    STATE_RUNNING,
    STATE_BLOCKED,
    STATE_PASSED,
    STATE_FAILED,
    STATE_SIGNING,
    STATE_ARCHIVED,
)

STATE_SCOPE_NODE: Final[StateScope] = "NODE"
STATE_SCOPE_CONTAINER: Final[StateScope] = "CONTAINER"

ERROR_UNKNOWN_SCOPE: Final[str] = "SM_STATE_UNKNOWN_SCOPE"
ERROR_UNKNOWN_STATUS: Final[str] = "SM_STATE_UNKNOWN_STATUS"
ERROR_ILLEGAL_TRANSITION: Final[str] = "SM_STATE_ILLEGAL_TRANSITION"

ERROR_DEFINITIONS: Final[dict[str, str]] = {
    ERROR_UNKNOWN_SCOPE: "State machine scope is invalid.",
    ERROR_UNKNOWN_STATUS: "State value is unknown and cannot be mapped.",
    ERROR_ILLEGAL_TRANSITION: "State transition is not allowed.",
}

STATE_TEXT_MAP_ZH_CN: Final[dict[UnifiedState, str]] = {
    STATE_INIT: "草稿",
    STATE_READY: "就绪",
    STATE_RUNNING: "执行中",
    STATE_BLOCKED: "已阻断",
    STATE_PASSED: "通过",
    STATE_FAILED: "不通过",
    STATE_SIGNING: "签名中",
    STATE_ARCHIVED: "已归档",
}

_STATE_ALIASES: Final[dict[str, UnifiedState]] = {
    "INIT": STATE_INIT,
    "READY": STATE_READY,
    "RUNNING": STATE_RUNNING,
    "BLOCKED": STATE_BLOCKED,
    "PASSED": STATE_PASSED,
    "FAILED": STATE_FAILED,
    "SIGNING": STATE_SIGNING,
    "ARCHIVED": STATE_ARCHIVED,
    # Legacy backend/frontend states.
    "DRAFT": STATE_INIT,
    "PENDING": STATE_INIT,
    "UNLOCKED": STATE_READY,
    "IN_PROGRESS": STATE_RUNNING,
    "COMPUTED": STATE_RUNNING,
    "GATED": STATE_RUNNING,
    "PASS": STATE_PASSED,
    "FINAL_PASS": STATE_PASSED,
    "QUALIFIED": STATE_PASSED,
    "VALIDATED": STATE_PASSED,
    "VERIFIED": STATE_PASSED,
    "SUCCESS": STATE_PASSED,
    "OVERRIDDEN": STATE_PASSED,
    "FAIL": STATE_FAILED,
    "FINAL_FAIL": STATE_FAILED,
    "REJECTED": STATE_FAILED,
    "CRITICAL": STATE_FAILED,
    "ERROR": STATE_FAILED,
    "LOCKED": STATE_BLOCKED,
}

_PAGE_TEXT_ALIASES: Final[dict[str, UnifiedState]] = {
    "草稿": STATE_INIT,
    "就绪": STATE_READY,
    "可执行": STATE_READY,
    "执行中": STATE_RUNNING,
    "进行中": STATE_RUNNING,
    "阻塞": STATE_BLOCKED,
    "受阻": STATE_BLOCKED,
    "已阻断": STATE_BLOCKED,
    "通过": STATE_PASSED,
    "已完成": STATE_PASSED,
    "已验证": STATE_PASSED,
    "合格": STATE_PASSED,
    "不通过": STATE_FAILED,
    "未通过": STATE_FAILED,
    "已驳回": STATE_FAILED,
    "签名中": STATE_SIGNING,
    "已归档": STATE_ARCHIVED,
}

_NODE_TRANSITIONS: Final[dict[UnifiedState, tuple[UnifiedState, ...]]] = {
    STATE_INIT: (STATE_READY, STATE_RUNNING, STATE_BLOCKED, STATE_PASSED, STATE_FAILED, STATE_SIGNING, STATE_ARCHIVED),
    STATE_READY: (STATE_RUNNING, STATE_BLOCKED, STATE_ARCHIVED),
    STATE_RUNNING: (STATE_PASSED, STATE_FAILED, STATE_BLOCKED, STATE_SIGNING, STATE_ARCHIVED),
    STATE_BLOCKED: (STATE_READY, STATE_RUNNING, STATE_FAILED, STATE_ARCHIVED),
    STATE_PASSED: (STATE_SIGNING, STATE_RUNNING, STATE_ARCHIVED),
    STATE_FAILED: (STATE_READY, STATE_RUNNING, STATE_SIGNING, STATE_ARCHIVED),
    STATE_SIGNING: (STATE_PASSED, STATE_FAILED, STATE_ARCHIVED),
    STATE_ARCHIVED: (),
}

_CONTAINER_TRANSITIONS: Final[dict[UnifiedState, tuple[UnifiedState, ...]]] = {
    STATE_INIT: (STATE_READY, STATE_RUNNING, STATE_BLOCKED),
    STATE_READY: (STATE_RUNNING, STATE_BLOCKED, STATE_ARCHIVED),
    STATE_RUNNING: (STATE_READY, STATE_BLOCKED, STATE_PASSED, STATE_FAILED, STATE_SIGNING, STATE_ARCHIVED),
    STATE_BLOCKED: (STATE_READY, STATE_RUNNING, STATE_FAILED, STATE_ARCHIVED),
    STATE_PASSED: (STATE_RUNNING, STATE_SIGNING, STATE_ARCHIVED),
    STATE_FAILED: (STATE_READY, STATE_RUNNING, STATE_SIGNING, STATE_ARCHIVED),
    STATE_SIGNING: (STATE_PASSED, STATE_FAILED, STATE_ARCHIVED),
    STATE_ARCHIVED: (),
}

_TRANSITIONS_BY_SCOPE: Final[dict[StateScope, dict[UnifiedState, tuple[UnifiedState, ...]]]] = {
    STATE_SCOPE_NODE: _NODE_TRANSITIONS,
    STATE_SCOPE_CONTAINER: _CONTAINER_TRANSITIONS,
}


@dataclass(frozen=True)
class StateTransitionError(ValueError):
    code: str
    message: str
    scope: str | None = None
    current: str | None = None
    target: str | None = None
    allowed: tuple[str, ...] = ()

    def __str__(self) -> str:  # pragma: no cover - trivial
        return self.message

    def as_dict(self) -> dict[str, object]:
        return {
            "code": self.code,
            "message": self.message,
            "scope": self.scope,
            "current": self.current,
            "target": self.target,
            "allowed": list(self.allowed),
        }


def _normalize_token(value: str | None) -> str:
    return str(value or "").strip().upper()


def _normalize_scope(scope: StateScope | str) -> StateScope:
    raw = _normalize_token(scope)
    if raw == STATE_SCOPE_NODE:
        return STATE_SCOPE_NODE
    if raw == STATE_SCOPE_CONTAINER:
        return STATE_SCOPE_CONTAINER
    raise StateTransitionError(
        code=ERROR_UNKNOWN_SCOPE,
        message=f"{ERROR_DEFINITIONS[ERROR_UNKNOWN_SCOPE]} scope={scope!r}",
        scope=str(scope),
    )


def normalize_state(value: str | None, *, scope: StateScope | str | None = None) -> UnifiedState:
    if scope is not None:
        _normalize_scope(scope)
    normalized = _normalize_token(value)
    mapped = _STATE_ALIASES.get(normalized)
    if mapped is not None:
        return mapped
    raise StateTransitionError(
        code=ERROR_UNKNOWN_STATUS,
        message=f"{ERROR_DEFINITIONS[ERROR_UNKNOWN_STATUS]} status={value!r}",
        scope=None if scope is None else _normalize_token(str(scope)),
        current=str(value),
    )


def normalize_page_status_text(value: str | None) -> UnifiedState:
    normalized = str(value or "").strip()
    if not normalized:
        raise StateTransitionError(
            code=ERROR_UNKNOWN_STATUS,
            message=f"{ERROR_DEFINITIONS[ERROR_UNKNOWN_STATUS]} status_text={value!r}",
        )
    mapped = _PAGE_TEXT_ALIASES.get(normalized)
    if mapped is not None:
        return mapped
    return normalize_state(normalized)


def allowed_targets(scope: StateScope | str, current_state: str | None) -> tuple[UnifiedState, ...]:
    normalized_scope = _normalize_scope(scope)
    normalized_current = normalize_state(current_state, scope=normalized_scope)
    transitions = _TRANSITIONS_BY_SCOPE[normalized_scope]
    return transitions[normalized_current]


def can_transition(scope: StateScope | str, current_state: str | None, target_state: str | None) -> bool:
    normalized_scope = _normalize_scope(scope)
    normalized_current = normalize_state(current_state, scope=normalized_scope)
    normalized_target = normalize_state(target_state, scope=normalized_scope)
    if normalized_current == normalized_target:
        return True
    return normalized_target in _TRANSITIONS_BY_SCOPE[normalized_scope][normalized_current]


def assert_transition(scope: StateScope | str, current_state: str | None, target_state: str | None) -> UnifiedState:
    normalized_scope = _normalize_scope(scope)
    normalized_current = normalize_state(current_state, scope=normalized_scope)
    normalized_target = normalize_state(target_state, scope=normalized_scope)
    if normalized_current == normalized_target:
        return normalized_target
    allowed = _TRANSITIONS_BY_SCOPE[normalized_scope][normalized_current]
    if normalized_target not in allowed:
        raise StateTransitionError(
            code=ERROR_ILLEGAL_TRANSITION,
            message=(
                f"{ERROR_DEFINITIONS[ERROR_ILLEGAL_TRANSITION]} "
                f"{normalized_scope}:{normalized_current}->{normalized_target}"
            ),
            scope=normalized_scope,
            current=normalized_current,
            target=normalized_target,
            allowed=tuple(allowed),
        )
    return normalized_target


def transition(scope: StateScope | str, current_state: str | None, target_state: str | None) -> UnifiedState:
    return assert_transition(scope=scope, current_state=current_state, target_state=target_state)
